### PR TITLE
ts/indicators-ichimoku-kinko-hyo

### LINKED ISSUE #11846: Migrate `indicators/ichimoku-kinko-hyo` to TypeScript

### PR DESCRIPTION
**Pull request by activist**
Closes #11846, Tools: Migrated IKHIndicator to TypeScript.

### CODE REVIEW

File: ts/indicators/ichimoku-kinko-hyo.src.ts [hunk 1]
```diff
@@ -0,0 +1,1009 @@
...
+                        gapSize: mainLineOptions.gapSize
+                    }
+                },
+                pointArrayMapLength: number = indicator.pointArrayMap.length,
+                allIchimokuPoints:
+                Array<Array<Highcharts.IKHIndicatorPoint>> =
+                [[], [], [], [], [], []],
+                ikhMap: Highcharts.Dictionary<
+                Array<Highcharts.IKHIndicatorPoint>
+                > = {
```
Comment (activist) [inline ts/indicators/ichimoku-kinko-hyo.src.ts]: The 80 char line limit kind of obscures the whole code some places in this file.

Comment (bre1470) [reply ts/indicators/ichimoku-kinko-hyo.src.ts]: Please use parentheses for multiline typing (see `ts/README.md`). In that case
```ts
ikhMap: Highcharts.Dictionary<(
    Array<Highcharts.IKHIndicatorPoint>
)> = {
```

File: samples/unit-tests/axis/label-ellipsis/demo.js [hunk 2]
```diff
@@ -69,7 +69,7 @@ QUnit.test('#5034: No ellipsis for multiline labels where there is room', functi
 
         chart: {
             type: "bar",
-            width: 500
+            width: 450
```
Comment (bre1470) [inline samples/unit-tests/axis/label-ellipsis/demo.js]: How can this be necessary?

File: ts/indicators/ichimoku-kinko-hyo.src.ts [hunk 3]
```diff
@@ -0,0 +1,1009 @@
+/* *
+ *
+ *  License: www.highcharts.com/license
+ *
```
Comment (bre1470) [inline ts/indicators/ichimoku-kinko-hyo.src.ts]: Warning about editing transpiled files is missing.

File: ts/indicators/ichimoku-kinko-hyo.src.ts [hunk 4]
```diff
@@ -0,0 +1,1009 @@
...
+            graph: (SVGElement|undefined);
+        }
+
+        interface IKHIndicatorSenkouSpanOptions {
+            color?: ColorType;
+            negativeColor?: ColorType;
+            styles?: CSSObject & { fill: ColorType };
+        }
+
+        interface IKHIndicatorGapExtension {
```
Comment (bre1470) [inline ts/indicators/ichimoku-kinko-hyo.src.ts]: This must end with `Object` name not with `Extension`.

File: ts/indicators/ichimoku-kinko-hyo.src.ts [hunk 5]
```diff
@@ -0,0 +1,1009 @@
...
+}
+
+function getClosestPointRange(axis: Highcharts.Axis): (number|undefined) {
+    var closestDataRange: (number|undefined),
+        loopLength: number,
+        distance: number,
+        xData: Array<number>,
+        i: number;
+
+    axis.series.forEach(function (series): void {
```
Comment (bre1470) [inline ts/indicators/ichimoku-kinko-hyo.src.ts]: Type for series parameter is missing.

File: ts/indicators/indicators.src.ts [hunk 6]
```diff
@@ -67,6 +67,12 @@ declare global {
             yData: Array<Array<number>>;
         }
 
+        interface IndicatorUndefinableValuesObject {
```
Comment (bre1470) [inline ts/indicators/indicators.src.ts]: Naming is incomplete as it is a `IndicatorMultipleUndefinableValuesObject`.

File: samples/unit-tests/axis/label-ellipsis/demo.js (hunk 2 above)
Comment (activist) [reply samples/unit-tests/axis/label-ellipsis/demo.js]: This is not related to the TS conversion, but after pulling latest changes from master this was needed as this test fails on Chrome/Linux. Thus, I could push without fixing the test as well.

File: ts/indicators/ichimoku-kinko-hyo.src.ts [hunk 7]
```diff
@@ -0,0 +1,1009 @@
...
+                yVal: Array<Array<number>> = (series.yData as any),
+                xAxis: Highcharts.Axis = series.xAxis,
+                yValLen: number = (yVal && yVal.length) || 0,
+                closestPointRange: number = (
+                    getClosestPointRange(xAxis) as any
+                ),
+                IKH: Array<Array<(number|undefined)>> = [],
+                xData: Array<number> = [],
+                dateStart: number,
+                date: (number|undefined),
```
Comment (bre1470) [inline ts/indicators/ichimoku-kinko-hyo.src.ts]: Try `date` with `number` only, as it is mandatory set in the for loop with the xVals. If this is working, you can remove all `date as any`.

File: ts/indicators/ichimoku-kinko-hyo.src.ts [hunk 8]
```diff
@@ -0,0 +1,1009 @@
...
+            public pointClass: typeof IKHIndicatorPoint;
+            public points: Array<IKHIndicatorPoint>;
+            public drawGraph(): void;
+            public init(): void;
+            public toYData(point: Point): Array<number>;
+            public translate(): void;
+
+        }
+
+        interface IKHIndicatorDrawSenkouSpanOptions {
```
Comment (bre1470) [inline ts/indicators/ichimoku-kinko-hyo.src.ts]: These are only arguments for a function, so the naming should be end on `Object`.

File: ts/indicators/ichimoku-kinko-hyo.src.ts (hunk 7 above)
Comment (activist) [reply ts/indicators/ichimoku-kinko-hyo.src.ts]: Unfortunately then the compiler reports `Variable 'date' is used before being assigned.` in the for-loop that is just before the return statement.

Comment (bre1470) [reply ts/indicators/ichimoku-kinko-hyo.src.ts]: Okay. Therefor the type is good as it is. 👍

**Comment by activist** [comment]
@bre1470 The requested changes has been added :)

File: ts/indicators/ichimoku-kinko-hyo.src.ts (hunk 8 above)
Comment (bre1470) [reply ts/indicators/ichimoku-kinko-hyo.src.ts]: This is not solved yet.

Comment (activist) [reply ts/indicators/ichimoku-kinko-hyo.src.ts]: Sorry, I could have swore that I renamed it. Corrected the name now.

File: ts/indicators/ichimoku-kinko-hyo.src.ts [hunk 9]
```diff
@@ -0,0 +1,1024 @@
...
+            indicator: IKHIndicator;
+            points: Array<IKHIndicatorPoint>;
+            nextPoints: Array<IKHIndicatorPoint>;
+            color?: ColorType;
+            options: IKHIndicatorOptions;
+            gap: IKHIndicatorGapExtensionObject;
+            graph: (SVGElement|undefined);
+        }
+
+        interface IKHIndicatorSenkouSpanObject {
```
Comment (bre1470) [inline ts/indicators/ichimoku-kinko-hyo.src.ts]: I guess you renamed the wrong interface yesterday. This one needs to end on `Options` (or `OptionsObject`) as it becomes part of the options tree in `IKHIndicatorOptions`.

Comment (activist) [reply ts/indicators/ichimoku-kinko-hyo.src.ts]: There you go. That is probably the culprit yes. Trying again with `IKHIndicatorSenkouSpanOptions`

**Pull request merged**